### PR TITLE
Adding tests for RunningJobs class

### PR DESCRIPTION
**Pull request by ironcerocloudbees**
## Add comprehensive test coverage for RunningJobs build cancellation

This PR adds complete test coverage for the `RunningJobs` class and related build cancellation functionality, which previously had 0% test coverage.

**What's included:**
- **17 unit tests** (`RunningJobsTest`) covering cancellation logic, policy validation, and edge cases
- **6 integration tests** (`RunningJobsHudsonTest`) verifying actual build cancellation in Jenkins queue and executors
- **10 tests** for interruption causes (`NewPatchSetInterruptionTest`, `AbandonedPatchsetInterruptionTest`)

**Test coverage:**
- ✅ New patchset cancellation (default behavior)
- ✅ Abort any patchset policy (including older patchsets)
- ✅ Queued build cancellation
- ✅ Abandoned patchset cancellation
- ✅ Topic-based cancellation
- ✅ Manual patchset handling
- ✅ Policy disabled scenarios
- ✅ Different changes don't interfere

This establishes baseline test coverage before the planned RunningJobs/BuildMemory unification.

### Testing done

**All tests passing:**
Tests run: 33, Failures: 0, Errors: 0, Skipped: 0

- Unit tests use Mockito to verify cancellation logic without Jenkins infrastructure
- Integration tests use `JenkinsRule` and `SshdServerMock` to verify actual build cancellation behavior
- All tests follow checkstyle rules and are fully documented

**Manual verification:**
- Tested all cancellation policies (enabled/disabled, various combinations)
- Verified builds are properly cancelled in queue and while running
- Confirmed different changes don't interfere with each other
- Validated interruption causes appear correctly in build logs

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests that demonstrate the feature works or the issue is fixed

### CODE REVIEW

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package com.sonyericsson.hudson.plugins.gerrit.trigger.hudsontrigger;
```
Comment (sboardwell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: Do we really need this test? It seems to be testing the obvious.

Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: Or is it just to increase the coverage?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package com.sonyericsson.hudson.plugins.gerrit.trigger.hudsontrigger;
```
Comment (sboardwell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: Same here.

Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: Or perhaps merge them all into one test?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,455 @@
+package com.sonyericsson.hudson.plugins.gerrit.trigger.hudsontrigger;
```
Comment (sboardwell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java:1]: Hudson in the name means we are talking about integration tests?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+            }
+            Thread.sleep(SLEEP_HALF_SEC);
+        }
+    }
+
+    /**
+     * Tests that a new patchset cancels a running build of an old patchset.
+     * This is the most common cancellation scenario.
+     *
+     * @throws Exception if test fails
```
Comment (sboardwell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Is this correct?

Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: This goes for all "if test fails" comments

**Review by sboardwell — COMMENTED** [review]
Looks good. Added some comments/questions.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java (hunk 1 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: Nah, it is true that those kinds of tests are useful. But it is indeed to increase coverage.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java (hunk 2 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: The same, it is for coverage. We should not merge the Junit test classes, right?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java (hunk 3 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java:1]: Yes, those are integration tests. I wanted a way to confirm that the logic (business logic) was working correctly (not only Junit tests to confirm that we are adding or removing elements from a collection).

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 4 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: I think it is correct. I have just one doubt:

Should the `abortNewPatchsets` policy setting be set to false? According to the code, it should be false, and, indeed, the test passed with this configuration. However, I haven't completed cleaning the concept of "new patchsets". To me, a new patchset is the same changeId with a different PatchSetNumber, right?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java (hunk 2 above)
Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: I meant merging the asserts

Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: The `testInstantiation` seems a little overkill. It will never fail and is therefore just noise (which could be said for other tests). Perhaps just add a single test which contains all the asserts. Still noise, but only one test.

Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: It's ok to remove the `testInstantiation` tests. However, I would like to keep the rest of the tests split (like `testIsCauseOfInterruption`, `testGetShortDescription`, etc); that will keep the concept of one Junit test for each feature/method/field.

**Comment by ironcerocloudbees** [comment]
@sboardwell, I renamed the integration test to better match once we migrate the logic from running jobs to build memory.

@sboardwell , @rsandell. I would like to proceed with this PR before opening the new PR for the migration. Can we go ahead?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java [hunk 5]
```diff
@@ -33,12 +33,12 @@
...
 
 /**
- * Integration tests for {@link RunningJobs} that verify actual job cancellation
+ * Integration tests for build cancellation feature that verify actual job cancellation
  * in Jenkins queue and running executors.
  *
  * @author Ignacio Roncero &lt;ironcero@cloudbees.com&gt;
  */
-public class RunningJobsHudsonTest {
+public class BuildCancellationHudsonTest {
```
Comment (sboardwell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Looks good - one question: Is `...HudsonTest` generally used in plugins, or would `...IntegrationTest` be more appropriate? I've never heard it before hence the confusion.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 4 above)
Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Sorry, misunderstanding. Throwing an exception is generally an error in the code in my book.

If you expect the code you are calling to fail and throw an exception, this should be handled. If the test itself throws an exception, this would be an unexpected error, no?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 5 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Ok, let me change it.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 4 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: I guess you are referring to the exception that the `waitForBuildToStart` method can throw, which the `testNewPatchsetCancelsRunningBuild` method at line 151 will not handle, right?
If so, I believe it should not be handled for the following reason:
Any exception in the `waitForBuildToStart` method is due to a system malfunction (the job did not start, or we reached the assigned timeout). If this exception occurs, the result of the test (`testNewPatchsetCancelsRunningBuild`) should not be a failing assertion; it should be a general exception that will completely break the tests' Maven goal to highlight the developer's action. So, the exception will travel from `waitForBuildToStart` to the Maven goal thanks to `throws Exception` in `testNewPatchsetCancelsRunningBuild`.

Is the above what you are asking for? If you think that we should handle it (in any case), I can modify it.

Comment (sboardwell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Okay, yes, then it is an unexpected error. The `if test fails` implies the `testNewPatchsetCancelsRunningBuild` test failed and not `waitForBuildToStart` method throwing an unexpected error.

Sorry, being pedantic but it doesn't read correctly.

Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: Ah, I get your point. The message in the Javadoc is not clear. You're right. Let me correct it.

**Review by sboardwell — APPROVED** [review]
LGTM 🚀

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java (hunk 1 above)
Comment (rsandell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: yea, these just seems to be wasting preacous build time. :)

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java (hunk 3 above)
Comment (rsandell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java:1]: The naming convention that I've used before in this plugin is ´*HudsonTest´ for the tests that are legacy JUnit3 tests that ´extends HudsonTestCase´ (yes the plugin is that old). Some of those might have been migrated to ´JenkinsRule´ but missed getting renamed in the process. So caling it ´*IntegrationTest´ or ´*JenkinsTest´ is fine.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+
+        // Trigger patchset 1 again (will be queued due to single executor)
+        gerritServer.triggerEvent(patchset1);
+
+        // Wait a bit for it to enter queue
+        Thread.sleep(SLEEP_HALF_SEC);
+
+        // Verify there's a build in queue
+        Queue.Item[] items = jenkins.jenkins.getQueue().getItems();
+        assertTrue("Should have queued build", items.length > 0);
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java]: Could have been written a bit nicer like


```suggestion
        await().until(() -> jenkins.jenkins.getQueue().getItems(), hasSize(greaterThan(0)))
```
But it is fine as is.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/RunningJobsTest.java [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link RunningJobs}.
+ *
+ * @author Ignacio Roncero &lt;ironcero@cloudbees.com&gt;
+ */
+@RunWith(MockitoJUnitRunner.Silent.class)
+public class RunningJobsTest {
+
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/RunningJobsTest.java]: I don't understand what these are supposed to be testing? That removing something from a standard java collection returns true?

File: src/test/resources/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest/common/gerrit-trigger.xml [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (rsandell) [inline src/test/resources/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest/common/gerrit-trigger.xml:1]: the config files looks to be the same, so I believe they could be just one file in ´src/test/resources/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest/´

It also looks like the default settings mostly so it could be added programattically perhaps? Some tests needs the configuration to be there during startup so that jenkins connects before the test can start, but that might not have relevance here?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Tests for {@link NewPatchSetInterruption}.
+ *
+ * @author Ignacio Roncero &lt;ironcero@cloudbees.com&gt;
+ */
+public class NewPatchSetInterruptionTest {
+
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: These looks irrelevant.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+        patchset2.getPatchSet().setNumber("2");
+
+        gerritServer.triggerEvent(patchset2);
+
+        // Wait for build1 to be aborted
+        jenkins.waitUntilNoActivity();
+
+        // Verify build1 was aborted
+        assertEquals(Result.ABORTED, build1.getResult());
+
```
Comment (rsandell) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java:184]: verify the cause of abortion?

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java (hunk 1 above)
Comment (rsandell) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: You can get most of the code coverage by verifying the cause of interruption in the integration tests ;)

**Review by rsandell — CHANGES_REQUESTED** [review]
The irrelevant tests should be removed imo, the plugin has problem fitting into the alloted build time on ci.jenkins.io as it is 🤓

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java (hunk 1 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/AbandonedPatchsetInterruptionTest.java]: Done, removed the AbandonedPatchsetInterruptionTest.java file and checking the cause of interruption in the integration tests added

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java (hunk 2 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: All done, resolving the conversation.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 4 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: All done, resolving the conversation

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java (hunk 5 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationHudsonTest.java]: All done, resolving the conversation

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/RunningJobsTest.java (hunk 7 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/RunningJobsTest.java]: Removed the RunningJobsTest.java file.

File: src/test/resources/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest/common/gerrit-trigger.xml (hunk 8 above)
Comment (ironcerocloudbees) [reply src/test/resources/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest/common/gerrit-trigger.xml:1]: Done (collapsing all the XML files into only one).

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java (hunk 9 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/NewPatchSetInterruptionTest.java]: Removed NewPatchsetInterruptionTest.java file.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java (hunk 10 above)
Comment (ironcerocloudbees) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/BuildCancellationIntegrationTest.java:184]: Verified in the integration tests now.

**Review by sboardwell — APPROVED** [review]
LGTM
